bugzooka/integrations/gemini_client.py

```diff
@@ -119,17 +184,102 @@ def analyze_log_with_gemini(
+                       [t["function"]["name"] for t in openai_tools])
+
+        # Tool calling loop - iterate until we get a final answer or hit max iterations
+        iteration = 0
+        while iteration < max_iterations:
+            iteration += 1
+
+            # Call Gemini API
+            api_kwargs = {
+                "messages": messages,
+                "model": model,
+                "max_tokens": INFERENCE_MAX_TOKENS,
+                "temperature": INFERENCE_TEMPERATURE,
+            }
+
+            # Only add tools if they exist
+            if openai_tools:
+                api_kwargs["tools"] = openai_tools
+
+            response = gemini_client.chat_completions_create(**api_kwargs)
+
+            response_message = response.choices[0].message
 
-        return response.choices[0].message.content
+            # Check if Gemini wants to call tools
+            tool_calls = getattr(response_message, 'tool_calls', None)
+
+            if not tool_calls:
+                # No tool calls - we have the final answer
+                logger.info("Gemini analysis complete (iteration %d)", iteration)
+                content = response_message.content
+                if content is None:
+                    logger.warning("Gemini returned None content, using empty string")
+                    return ""
+                return content
+
+            # Gemini wants to call tools - execute them
+            logger.info("Gemini requested %d tool call(s)", len(tool_calls))
+
+            # Add the assistant's message with tool calls to conversation
+            messages.append({
+                "role": "assistant",
+                "content": response_message.content or "",
+                "tool_calls": [
+                    {
+                        "id": tc.id,
+                        "type": "function",
+                        "function": {
+                            "name": tc.function.name,
+                            "arguments": tc.function.arguments
+                        }
+                    }
+                    for tc in tool_calls
+                ]
+            })
+
+            # Execute each tool call and add results to messages
+            for tool_call in tool_calls:
+                function_name = tool_call.function.name
+                try:
+                    function_args = json.loads(tool_call.function.arguments)
+                except json.JSONDecodeError as e:
+                    logger.error("Failed to parse tool arguments: %s", e)
+                    function_result = f"Error: Invalid JSON arguments - {str(e)}"
+                else:
+                    # Execute the tool (await since it's now async)
+                    function_result = await execute_tool_call(
+                        function_name,
+                        function_args,
+                        tools
+                    )
+
+                # Add tool result to messages
+                messages.append({
```

Comment:
Its good that we are capturing these messages, but are we logging them somewhere? or is it just solely for the testing purposes?

Reply:
They're currently not logged but they're captured in the messages list for functional purposes as they're used between iterations.